presto-main/src/main/java/com/facebook/presto/memory/HighMemoryTaskKiller.java

```diff
@@ -0,0 +1,248 @@
+                if (shouldTriggerTaskKiller(info)) {
+                    //Kill task consuming most memory
+                    List<SqlTask> activeTasks = getActiveTasks();
+                    ListMultimap<QueryId, SqlTask> activeQueriesToTasksMap = activeTasks.stream()
+                            .collect(toImmutableListMultimap(task -> task.getQueryContext().getQueryId(), Function.identity()));
+
+                    Optional<QueryId> queryId = getMaxMemoryConsumingQuery(activeQueriesToTasksMap);
+
+                    if (queryId.isPresent()) {
+                        List<SqlTask> activeTasksToKill = activeQueriesToTasksMap.get(queryId.get());
+                        for (SqlTask sqlTask : activeTasksToKill) {
+                            TaskStats taskStats = sqlTask.getTaskInfo().getStats();
+                            sqlTask.failed(new PrestoException(EXCEEDED_HEAP_MEMORY_LIMIT, format("Worker heap memory limit exceeded: User Memory: %d, System Memory: %d, Revocable Memory: %d", taskStats.getUserMemoryReservationInBytes(), taskStats.getSystemMemoryReservationInBytes(), taskStats.getRevocableMemoryReservationInBytes())));
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private boolean shouldTriggerTaskKiller(GarbageCollectionNotificationInfo info)
+    {
+        boolean triggerTaskKiller = false;
+        DataSize beforeGcDataSize = info.getBeforeGcTotal();
+        DataSize afterGcDataSize = info.getAfterGcTotal();
+
+        if (taskKillerStrategy == FREE_MEMORY_ON_FREQUENT_FULL_GC) {
+            long currentGarbageCollectedBytes = beforeGcDataSize.toBytes() - afterGcDataSize.toBytes();
+            long currentFullGCTimestamp = ticker.read();
+
+            if (isFrequentFullGC(lastFullGCTimestamp, currentFullGCTimestamp) && !hasFullGCFreedEnoughBytes(currentGarbageCollectedBytes)) {
+                triggerTaskKiller = true;
+            }
+
+            lastFullGCTimestamp = currentFullGCTimestamp;
+            lastFullGCCollectedBytes = currentGarbageCollectedBytes;
+        }
+        else if (taskKillerStrategy == FREE_MEMORY_ON_FULL_GC) {
+            if (isLowMemory() && beforeGcDataSize.toBytes() - afterGcDataSize.toBytes() < reclaimMemoryThreshold) {
+                triggerTaskKiller = true;
+            }
+        }
+
+        return triggerTaskKiller;
+    }
+
+    private List<SqlTask> getActiveTasks()
+    {
+        return sqlTaskManager.getAllTasks().stream()
+                .filter(task -> !task.getTaskState().isDone())
+                .collect(toImmutableList());
+    }
+
+    @VisibleForTesting
+    public static Optional<QueryId> getMaxMemoryConsumingQuery(ListMultimap<QueryId, SqlTask> queryIDToSqlTaskMap)
+    {
+        if (queryIDToSqlTaskMap.isEmpty()) {
+            return Optional.empty();
+        }
+
+        Comparator<Map.Entry<QueryId, Long>> comparator = Comparator.comparingLong(Map.Entry::getValue);
```

Comment:
Why not put the .reversed here ?

Reply:
due to having static vs non static method, this is not straight moving the reversed at the top so going to address in followup PR.